template/express-rest-api/Dockerfile

```diff
@@ -0,0 +1,67 @@
+# Docker image history includes ARG values, so never target this stage directly
+FROM node:12-alpine AS unsafe-deps
+
+WORKDIR /workdir
+
+COPY package.json yarn.lock ./
+
+ARG NPM_READ_TOKEN
+
+RUN \
+  echo '//registry.npmjs.org/:_authToken=${NPM_READ_TOKEN}' > .npmrc && \
+  yarn install --frozen-lockfile --ignore-optional --non-interactive --production && \
+  rm .npmrc
+
+###
+
+# Docker image history includes ARG values, so never target this stage directly
+FROM unsafe-deps AS unsafe-dev-deps
+
+ARG NPM_READ_TOKEN
+
+RUN \
+  echo '//registry.npmjs.org/:_authToken=${NPM_READ_TOKEN}' > .npmrc && \
+  yarn install --frozen-lockfile --ignore-optional --non-interactive && \
+  rm .npmrc
+
+###
+
+FROM node:12-alpine AS deps
+
+WORKDIR /workdir
+
+COPY --from=unsafe-deps /workdir .
+
+###
+
+FROM node:12-alpine AS dev-deps
+
+WORKDIR /workdir
+
+COPY --from=unsafe-dev-deps /workdir .
+
+###
+
+FROM dev-deps AS build
+
+COPY . .
+
+RUN yarn build
+
+###
+
+FROM node:12-alpine AS runtime
+
+WORKDIR /workdir
+
+COPY --from=build /workdir/lib lib
+
+COPY --from=deps /workdir/node_modules node_modules
+
+ENV NODE_ENV production
+
+ARG PORT=3000
+ENV PORT ${PORT}
+EXPOSE ${PORT}
+
+CMD node lib/main
```

Comment:
Is this what you run in production? Granted I wouldn't expect these frameworks to properly crash often, but spinning up a new ECS task would be slower than having a clustering implementation spawn another process.

Reply:
Yes, we are running our some of our project like this in production. Do you recommend running node with [pm2-runtime](https://pm2.keymetrics.io/docs/usage/docker-pm2-nodejs/) ?

Reply:
I think there's an open question for us to resolve around when multiple processes make sense with constrained compute in Fargate. I'm happy to leave this for now.

Reply:
Hi @72636c , we did use pm2 for one of our services in the past, but stopped. Main reason: we don't expect crashes to be a part of a normal application lifecycle. It also was causing us some issues, when app was keep crashing but we thought it was running (but probably because of the way, how we configured our monitors in DataDog).

> I think there's an open question for us to resolve around when multiple processes make sense with constrained compute in Fargate.

I see this question is more like an advanced topic, which requires to have proper circumstances. We don't have it yet. So I would agree with you that we might leave it out of scope for now.

p.s. on a personal note I would like to hear when and how you use pm2 to overcome "constrained compute in Fargate".